File: randomized_quick_sort.c
#include <stdio.h>
#include <stdlib.h> // For rand() and srand()
#include <time.h>   // For seeding the random number generator

void swap(int *p, int *q) {
    int t;
    t = *p;
    *p = *q;
    *q = t;
}

int randomPartition(int a[], int f, int l) {
    // Generate a random index between f and l
    int randomIndex = f + rand() % (l - f + 1);
    
    // Swap the randomly selected element with the first element
    swap(&a[f], &a[randomIndex]);

    int key = a[f], i = f + 1, j = l;

    do {
        while (i <= j && a[i] < key)
            i++;
        while (j >= f && a[j] > key)
            j--;
        if (i < j)
            swap(&a[i], &a[j]);
    } while (i <= j);

    swap(&a[j], &a[f]);
    return (j);
}

void randomQSort(int a[], int l, int r) {
    int m;
    if (l < r) {
        m = randomPartition(a, l, r);
        randomQSort(a, l, m - 1);
        randomQSort(a, m + 1, r);
    }
}

int main() {
    int n, i, a[100];

    printf("Enter N: ");
    scanf("%d", &n);

    srand(time(NULL)); // Seed the random number generator

    printf("Enter the elements to be sorted:\n");
    for (i = 0; i < n; i++)
        scanf("%d", &a[i]);

    randomQSort(a, 0, n - 1);

    printf("Sorted array:\n");
    for (i = 0; i < n; i++)
        printf("%d ", a[i]);

    return 0;
}

File: randomized_quicksort.py
import random

def random_partition(arr, f, l):
    random_index = random.randint(f, l)
    arr[f], arr[random_index] = arr[random_index], arr[f]

    key = arr[f]
    i, j = f + 1, l

    while True:
        while i <= j and arr[i] < key:
            i += 1
        while j >= f and arr[j] > key:
            j -= 1
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]
        else:
            break

    arr[j], arr[f] = arr[f], arr[j]
    return j

def random_qsort(arr, l, r):
    if l < r:
        m = random_partition(arr, l, r)
        random_qsort(arr, l, m - 1)
        random_qsort(arr, m + 1, r)

if __name__ == "__main__":
    n = int(input("Enter N: "))
    a = []

    print("Enter the elements to be sorted:")
    for i in range(n):
        a.append(int(input()))

    random_qsort(a, 0, n - 1)

    print("Sorted array:")
    for i in range(n):
        print(a[i], end=" ")
